src/pojo-interfaces.ts

```diff
@@ -0,0 +1,418 @@
+    type: "Subject",
+    url: subjectIri,
+    predicates: {},
+  };
+  return freeze({
+    ...graph,
+    [subjectIri]: addRdfJsQuadToSubject(subject, quad, quadParseOptions),
+  });
+}
+
+function addRdfJsQuadToSubject(
+  subject: Subject,
+  quad: RdfJs.Quad,
+  quadParseOptions: QuadParseOptions = {}
+): Subject {
+  return freeze({
+    ...subject,
+    predicates: addRdfJsQuadToPredicates(
+      subject.predicates,
+      quad,
+      quadParseOptions
+    ),
+  });
+}
+
+function addRdfJsQuadToPredicates(
+  predicates: Predicates,
+  quad: RdfJs.Quad,
+  quadParseOptions: QuadParseOptions = {}
+): Predicates {
+  const supportedPredicateTypes: Array<typeof quad.predicate.termType> = [
+    "NamedNode",
+  ];
+  if (!supportedPredicateTypes.includes(quad.predicate.termType)) {
+    throw new Error(
+      `Cannot parse Quads with nodes of type [${quad.predicate.termType}] as their Predicate node.`
+    );
+  }
+  const predicateIri = quad.predicate.value;
+  const objects = predicates[predicateIri] ?? {};
+  return freeze({
+    ...predicates,
+    [predicateIri]: addRdfJsQuadToObjects(objects, quad, quadParseOptions),
+  });
+}
+
+function addRdfJsQuadToObjects(
+  objects: Objects,
+  quad: RdfJs.Quad,
+  quadParseOptions: QuadParseOptions = {}
+): Objects {
+  if (quad.object.termType === "NamedNode") {
+    const namedNodes = freeze([
+      ...(objects.namedNodes ?? []),
+      quad.object.value,
+    ]);
+    return freeze({
+      ...objects,
+      namedNodes: namedNodes,
+    });
+  }
+
+  if (quad.object.termType === "Literal") {
+    if (quad.object.datatype.value === xmlSchemaTypes.langString) {
+      const locale = quad.object.language.toLowerCase();
+      const thisLocaleStrings = freeze([
+        ...(objects.langStrings?.[locale] ?? []),
+        quad.object.value,
+      ]);
+      const langStrings = freeze({
+        ...(objects.langStrings ?? {}),
+        [locale]: thisLocaleStrings,
+      });
+      return freeze({
+        ...objects,
+        langStrings: langStrings,
+      });
+    }
+
+    // If the Object is a non-langString Literal
+    const thisTypeValues = freeze([
+      ...(objects.literals?.[quad.object.datatype.value] ?? []),
+      quad.object.value,
+    ]);
+    const literals = freeze({
+      ...(objects.literals ?? {}),
+      [quad.object.datatype.value]: thisTypeValues,
+    });
+    return freeze({
+      ...objects,
+      literals: literals,
+    });
+  }
+
+  if (quad.object.termType === "BlankNode") {
+    const blankNodePredicates = getPredicatesForBlankNode(
+      quadParseOptions.otherQuads ?? [],
+      quad.object
+    );
+    const blankNodes = freeze([
+      ...(objects.blankNodes ?? []),
+      blankNodePredicates,
+    ]);
+    return freeze({
+      ...objects,
+      blankNodes: blankNodes,
+    });
+  }
+
+  throw new Error(
+    `Objects of type [${quad.object.termType}] are not supported.`
+  );
+}
+
+function addRdfJsBlankNodeObjectsToPredicates(
```

Comment:
The only concern I have with the current approach to blank nodes is that, if I understand correctly, blank nodes would be effectively duplicated each time they appear as an object. In some cases, that may result in fairly large objects being cloned and passed around. I suspect with native `Object` methods being used, the overhead will not be too consequent, so that may be fine. I don't really see a satisfactory way (as in, not too hacky) to give blank node an identifier that would be consistent from one serialization to the other to enable comparison anyway, so we can simply wait and see if that becomes an issue in the future.

Reply:
Yep, my guess here would be that it is uncommon to have many quads with the same blank node as the object in the use cases in which you'd use an immutable dataset (at that point it's easier, and I'd encourage, to reach for named nodes), so being able to deduplicate them across different instances and not having to use hacky workarounds outweighs that in my view.

To have a 1:1 correspondence the other way around (i.e. from an arbitrary RDF/JS Dataset to this one), we'd probably still at some point need arbitrary unstable names for blank nodes that only appear in either Subject or Object position but not both, or just reject them at point of import.

Either way there's nothing blocking any such approach in the future, so we can still change it if implementation experience teaches provides new insights.